modules/library/data/src/main/java/org/geotools/data/memory/MemoryFeatureReader.java

```diff
@@ -30,20 +34,27 @@
 import org.opengis.feature.simple.SimpleFeatureType;
 
 public class MemoryFeatureReader implements FeatureReader<SimpleFeatureType, SimpleFeature>{
-    ContentState state;
+
     SimpleFeatureType featureType;
     Iterator<SimpleFeature> iterator;
 
     public MemoryFeatureReader(ContentState state, Query query) throws IOException {
-        this.state = state;
         featureType = state.getFeatureType();
-        String typeName = getFeatureType().getTypeName();
-        MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
-        iterator = store.features(typeName).values().iterator();
+
+        final MemoryDataStore store = (MemoryDataStore) state.getEntry().getDataStore();
+        final Map<String, SimpleFeature> features = store.features(featureType.getTypeName());
+        synchronized (features) {
```

Comment:
I don't see corresponding synchronizations in the writer. It would seem this change makes the exception less likely, but not impossible?
What about using a Concurrent map also to store this map instead? Is the memory overhead too high?